### ISSUE TITLE
Improper resolution of referenced subschema

### ISSUE THREAD

**Issue by cblock** [jonasschmidt/ex_json_schema#41]
I got an `ExJsonSchema.Schema.InvalidReferenceError` when trying to resolve (Draft 04 compliant) [Signal-K json schema](https://signalk.org/specification/1.0.0/schemas/signalk.json#). 

To me it looks like the resolver incorrectly amends the `$ref": "aircraft.json#"` to the root `"id": "https://signalk.org/specification/1.0.0/schemas/signalk.json#` element instead of replacing the `signalk.json#` part in the uri with `aircraft.json#` (as described in [section 8.3.2 of the json schema spec](http://json-schema.org/latest/json-schema-core.html#rfc.section.8.3.2)).

Attached please find my local stackrace: 

```
** (ExJsonSchema.Schema.InvalidReferenceError) invalid reference https://signalk.org/specification/1.0.0/schemas/signalk.json#aircraft.json#
     stacktrace:
       (ex_json_schema) lib/ex_json_schema/schema.ex:287: ExJsonSchema.Schema.raise_invalid_reference_error/1
       (ex_json_schema) lib/ex_json_schema/schema.ex:150: ExJsonSchema.Schema.resolve_ref/2
       (ex_json_schema) lib/ex_json_schema/schema.ex:160: ExJsonSchema.Schema.resolve_ref!/2
       (ex_json_schema) lib/ex_json_schema/schema.ex:138: ExJsonSchema.Schema.resolve_property/3
       (ex_json_schema) lib/ex_json_schema/schema.ex:108: anonymous fn/3 in ExJsonSchema.Schema.do_resolve/3
       (stdlib) maps.erl:257: :maps.fold_1/3
       (ex_json_schema) lib/ex_json_schema/schema.ex:107: ExJsonSchema.Schema.do_resolve/3
       (ex_json_schema) lib/ex_json_schema/schema.ex:116: ExJsonSchema.Schema.resolve_property/3
       (ex_json_schema) lib/ex_json_schema/schema.ex:108: anonymous fn/3 in ExJsonSchema.Schema.do_resolve/3
       (stdlib) maps.erl:257: :maps.fold_1/3
       (ex_json_schema) lib/ex_json_schema/schema.ex:107: ExJsonSchema.Schema.do_resolve/3
       (ex_json_schema) lib/ex_json_schema/schema.ex:116: ExJsonSchema.Schema.resolve_property/3
       (ex_json_schema) lib/ex_json_schema/schema.ex:108: anonymous fn/3 in ExJsonSchema.Schema.do_resolve/3
       (stdlib) maps.erl:257: :maps.fold_1/3
       (ex_json_schema) lib/ex_json_schema/schema.ex:107: ExJsonSchema.Schema.do_resolve/3
       (ex_json_schema) lib/ex_json_schema/schema.ex:116: ExJsonSchema.Schema.resolve_property/3
       (ex_json_schema) lib/ex_json_schema/schema.ex:108: anonymous fn/3 in ExJsonSchema.Schema.do_resolve/3
       (stdlib) maps.erl:257: :maps.fold_1/3
       (ex_json_schema) lib/ex_json_schema/schema.ex:107: ExJsonSchema.Schema.do_resolve/3
       (ex_json_schema) lib/ex_json_schema/schema.ex:116: ExJsonSchema.Schema.resolve_property/3
```

**Comment by jonasschmidt** [comment]
Yes, that definitely looks wrong. I think the official test suite now has a lot more tests around relative refs and how they should be resolved. Back when I implemented the ref logic there wasn't much to go on. I will try and come up with a fix this week, and maybe update the test suite to cover all the expected ref behaviors.

**Comment by cblock** [comment]
Great, thanks a lot! I can offer to test any changes against my current project if that would be of any help for your.

**Comment by jonasschmidt** [comment]
Ok, that one was bit trickier to fix than I expected, but I found a whole bunch of other bugs around ref resolution in the process. I have to make a few final changes and then I will release a fixed version later this week.

**Comment by cblock** [comment]
Great, thanks a lot! Looking forward to the new version!

**Comment by jonasschmidt** [comment]
I released the ref handling fix as version `0.6.0`, but it turns out that the Signal-K schema isn't a valid draft 4 schema:
```elixir
** (ExJsonSchema.Schema.InvalidSchemaError) schema did not pass validation against its meta-schema: [%ExJsonSchema.Validator.Error{error: %ExJsonSchema.Validator.Error.Type{actual: "array", expected: ["object"]}, path: "#/definitions/chargerQualities/properties/allOf"}]
```
I verified with other validator implementations and they all come to the same result. The schema in https://signalk.org/specification/1.0.0/schemas/groups/electrical.json tries to use `allOf` in a property definition context and that doesn't work. According to the spec, `allOf` is treated as a normal property in that context and property definitions have to be objects.
